src/vcl/utils.rs

```diff
@@ -1,7 +1,66 @@
+use std::ffi::{c_char, CStr};
+use std::slice::from_raw_parts;
+use std::str::from_utf8;
+
 use crate::ffi;
 use crate::ffi::{vfp_entry, vrt_ctx};
 use crate::vcl::backend::VCLBackendPtr;
 
+impl ffi::txt {
+    pub fn from_bytes(s: &[u8]) -> Self {
+        Self {
+            b: s.as_ptr().cast::<c_char>(),
+            e: unsafe { s.as_ptr().add(s.len()).cast::<c_char>() },
+        }
+    }
+
+    /// FIXME: This method is only used when calling [`ffi::VSLbt`],
+    /// and current implementation creates a string without a null terminator to pass it in.
+    /// Going forward, we should probably refactor it to avoid extra string allocation.
+    #[allow(clippy::should_implement_trait)]
+    pub fn from_str(s: &str) -> Self {
+        Self::from_bytes(s.as_bytes())
+    }
+
+    pub fn from_cstr(s: &CStr) -> Self {
+        Self::from_bytes(s.to_bytes())
+    }
+
+    /// Convert the `txt` struct to a `&[u8]`.
+    /// FIXME: Consider modifying this to return an empty slice instead of None,
+    ///        Unless we want to explicitly differentiate between empty and null.
+    pub fn to_slice<'a>(&self) -> Option<&'a [u8]> {
+        if self.b.is_null() {
+            None
+        } else {
+            // SAFETY: We assume that txt instance was created correctly,
+            //         so the pointers are valid and the end is after the beginning.
+            //         Txt instances are part of ffi, so inherently unsafe.
+            unsafe {
+                Some(from_raw_parts(
+                    self.b.cast::<u8>(),
+                    self.e.offset_from(self.b) as usize,
+                ))
+            }
+        }
+    }
+
+    /// Convert the `txt` struct to a `&str`.  Will panic if the string is not valid UTF-8.
+    pub fn to_str<'a>(&self) -> Option<&'a str> {
+        self.to_slice().map(|s| from_utf8(s).unwrap())
+    }
+
+    /// Parse the `txt` struct as a header, returning a tuple with the key and value,
+    /// trimming the value of leading whitespace.
+    pub fn parse_header<'a>(&self) -> Option<(&'a str, &'a str)> {
+        // We expect varnishd to always given us a string with a ':' in it
+        // If it's not the case, blow up as it might be a sign of a bigger problem.
+        let (key, value) = self.to_str()?.split_once(':').unwrap();
+        // FIXME: Consider `.trim_ascii_start()` if unicode is not a concern
```

Comment:
header names will be ascii, so go for it

Reply:
header name is not the issue - `.trim_start` uses `char::is_whitespace`, whereas `.trim_ascii_start` only removes "ascii whitespace" by using `u8::is_ascii_whitespace`.  Which means that if a header **value** can have non-ascii whitespace characters in the front, they will stay there.  Let's merge and decide this separate?